// Protocol buffer for the ProGraML representation.

syntax = "proto2";

package phd.deeplearning.ml4pl;

// A program graph.
message ProgramGraph {
  required int64 bytecode_id = 1;
  repeated Node node = 2;
  repeated Edge edge = 3;
}

message Node {
  enum Type {
    STATEMENT = 1;
    IDENTIFIER = 2;
    IMMEDIATE = 3;
  }

  required Type type = 1;
  required string text = 2;
  required string function = 3;
}

message Edge {
  enum Type {
    CONTROL = 1;
    DATA = 2;
    CALL = 3;
  }

  required Type type = 1;
  required int32 position = 2 [default=0];
}
